Text expand port

Blackbaud-TrevorBurch opened the pull request:
#162

blackbaud-sky-savage commented:
**Tests failed.** Automated cross-browser testing via BrowserStack and Travis CI shows that the JavaScript changes in this pull request are: **BUSTED**

Commit: f36a37522b1de37f1d4e57ecfc92bbb8b3e659ca
Build details: https://travis-ci.org/blackbaud-sky-savage/skyux2/builds/211763885

(*Please note that this is a [fully automated](https://github.com/twbs/savage) comment.*)

Blackbaud-TrevorBurch commented:
I'll be looking into this. Looks like something went wrong on some of my new tests when using Firefox

blackbaud-sky-savage commented:
**Tests failed.** Automated cross-browser testing via BrowserStack and Travis CI shows that the JavaScript changes in this pull request are: **BUSTED**

Commit: 83124607f7f559318603ea3fa55a73cfca9c59aa
Build details: https://travis-ci.org/blackbaud-sky-savage/skyux2/builds/211799641

(*Please note that this is a [fully automated](https://github.com/twbs/savage) comment.*)

blackbaud-johnly commented on src/app/components/text-expand/index.html:
```diff
@@ -0,0 +1,53 @@
+      modal view with the content will be displayed when the 'See more' link is clicked
+    </sky-demo-page-property>
+    <sky-demo-page-property propertyName="maxExpandedNewlines" defaultValue="2" isOptional="true">
+      The maximum number of newline characters to show in the inline expanded view. If there are more newline characters in the
```
I'd change "there are more newline characters in the content" to "the text includes more newline characters"

blackbaud-sky-savage commented:
**Tests failed.** Automated cross-browser testing via BrowserStack and Travis CI shows that the JavaScript changes in this pull request are: **BUSTED**

Commit: 44d1b03381a0ed4c87a8170d433522a1b7da2fd9
Build details: https://travis-ci.org/blackbaud-sky-savage/skyux2/builds/211825811

(*Please note that this is a [fully automated](https://github.com/twbs/savage) comment.*)

blackbaud-johnly commented on src/app/components/text-expand/index.html:
```diff
@@ -0,0 +1,53 @@
+
+  <sky-demo-page-properties sectionHeading="Text expand repeater properties">
+    <sky-demo-page-property propertyName="data">
+      The data to truncate
```
Add a period after "truncate."

blackbaud-johnly commented on src/app/components/text-expand/index.html:
```diff
@@ -0,0 +1,53 @@
+      The data to truncate
+    </sky-demo-page-property>
+    <sky-demo-page-property propertyName="maxItems">
+      The number of items from the data to display before truncating
```
Add a period after "truncating."

Blackbaud-TrevorBurch commented:
@blackbaud-johnly I will work on those. I pretty much ported the Sky 1 docs so I would expect there to be the same issues.

blackbaud-sky-savage commented:
**Tests failed.** Automated cross-browser testing via BrowserStack and Travis CI shows that the JavaScript changes in this pull request are: **BUSTED**

Commit: deb1d5404625729b356cdbae95387e3cc3f92f36
Build details: https://travis-ci.org/blackbaud-sky-savage/skyux2/builds/211840049

(*Please note that this is a [fully automated](https://github.com/twbs/savage) comment.*)

Blackbaud-TrevorBurch commented:
@blackbaud-sky-savage retry

blackbaud-sky-savage commented:
**Tests failed.** Automated cross-browser testing via BrowserStack and Travis CI shows that the JavaScript changes in this pull request are: **BUSTED**

Commit: deb1d5404625729b356cdbae95387e3cc3f92f36
Build details: https://travis-ci.org/blackbaud-sky-savage/skyux2/builds/211850472

(*Please note that this is a [fully automated](https://github.com/twbs/savage) comment.*)

Blackbaud-TrevorBurch commented:
I have an automation test failing. I'm getting it to reproduce locally now so I'm looking into it. I'll get this eventually :)

blackbaud-sky-savage commented:
**Tests passed.** Automated cross-browser testing via BrowserStack and Travis CI shows that the JavaScript changes in this pull request are: **CONFIRMED**

Commit: e3434b0f9907ed7b26c9d7e9e8975376ed526a8e
Build details: https://travis-ci.org/blackbaud-sky-savage/skyux2/builds/211868073

(*Please note that this is a [fully automated](https://github.com/twbs/savage) comment.*)

blackbaud-sky-savage commented:
**Tests passed.** Automated cross-browser testing via BrowserStack and Travis CI shows that the JavaScript changes in this pull request are: **CONFIRMED**

Commit: e1f79d7caa700150ba71eff2a4ebb01b3b90e857
Build details: https://travis-ci.org/blackbaud-sky-savage/skyux2/builds/213501246

(*Please note that this is a [fully automated](https://github.com/twbs/savage) comment.*)

Blackbaud-PatrickOFriel commented on src/modules/text-expand-repeater/text-expand-repeater.component.ts:
```diff
@@ -0,0 +1,106 @@
+    if (this._data) {
+      this.items = this.elRef.nativeElement.querySelectorAll('.sky-text-expand-repeater-item');
+      for (let i = this.maxItems; i < this._data.length; i++) {
+        this.items[i].style.display = 'none';
```
Instead of manipulating the dom manually here, let's create an adapter service which does this for us. See https://github.com/blackbaud/skyux2/blob/master/src/modules/search/search-adapter.service.ts for an example of separating dom manipulation from our components. The reason for this is that in the future, when we use non-browser renderers, we can just swap out that service with new functionality.

Blackbaud-PatrickOFriel commented on src/modules/text-expand-repeater/text-expand-repeater.component.ts:
```diff
@@ -0,0 +1,106 @@
+    }
+    let component = this;
+    let container = <HTMLElement>this.containerEl.nativeElement;
+    container.addEventListener('transitionend',
```
Instead of doing this, we can use angular 2 animations. If you scroll to the bottom of [here](https://angular.io/docs/ts/latest/guide/animations.html), you can see with angular 2 you have the ability to do animation start and end events. This could be similar to how we did animations in the search component. https://github.com/blackbaud/skyux2/blob/master/src/modules/search/search.component.ts. In the start and done parts of the animation you can do whatever height bounding/style changes you need (with calls to an adapter like above please).

Blackbaud-PatrickOFriel commented on src/modules/text-expand-repeater/text-expand-repeater.component.visual-spec.js:
```diff
@@ -0,0 +1,23 @@
+      .setupTest('/text-expand-repeater.html')
+      .compareScreenshot({
+        screenshotName: 'text-expand-repeater-not-expanded',
+        selector: '#text-expand-repeater'
```
Add `checkAccessibility: true` to this to add our accessibility automation

Blackbaud-PatrickOFriel commented on src/modules/text-expand-repeater/text-expand-repeater.component.visual-spec.js:
```diff
@@ -0,0 +1,23 @@
+      .click('.sky-text-expand-see-more')
+      .compareScreenshot({
+        screenshotName: 'text-expand-repeater-expanded',
+        selector: '#text-expand-repeater'
```
Same as above

Blackbaud-PatrickOFriel commented on src/modules/text-expand/text-expand.component.html:
```diff
@@ -0,0 +1,5 @@
+<div class="sky-text-expand-container" #container>
+  <span class="sky-text-expand-text" #text></span>
+  <span class="sky-text-expand-ellipsis" *ngIf='!isExpanded && expandable'>... </span>
```
Use double quotes with this ngIf

Blackbaud-PatrickOFriel commented on src/modules/text-expand/text-expand.component.html:
```diff
@@ -0,0 +1,5 @@
+<div class="sky-text-expand-container" #container>
+  <span class="sky-text-expand-text" #text></span>
+  <span class="sky-text-expand-ellipsis" *ngIf='!isExpanded && expandable'>... </span>
+  <button type="button" class="sky-btn sky-btn-link-inline sky-text-expand-see-more" *ngIf='expandable' (click)="textExpand()">{{buttonText}}</button>
```
Same as above

Blackbaud-PatrickOFriel commented on src/modules/text-expand/text-expand.component.ts:
```diff
@@ -0,0 +1,182 @@
+  public ngAfterViewInit() {
+    let component = this;
+    let container = <HTMLElement>this.containerEl.nativeElement;
+    container.addEventListener('transitionend',
```
Let's try to do angular animations like I outlined with the repeater expand

Blackbaud-PatrickOFriel commented on src/modules/text-expand/text-expand.component.ts:
```diff
@@ -0,0 +1,182 @@
+
+  private animateText(previousText: string, newText: string,
+    expanding: boolean) {
+    let container = this.containerEl.nativeElement;
```
Let's try to move all dom manipulation to an adapter

Blackbaud-PatrickOFriel commented on src/modules/text-expand-repeater/text-expand-repeater.component.ts:
```diff
@@ -0,0 +1,106 @@
+  }
+
+  public repeaterExpand() {
+    let container = this.containerEl.nativeElement;
```
Let's again move all dom manipulation to an adapter

Blackbaud-TrevorBurch commented:
@Blackbaud-PatrickOFriel CI builds are running now on my changes. I was not able to get angular animations going and so I moved on as we discussed on Slack. I did move the binding of the transitionend event to the template instead of using the addEventListener method. I also fixed the accessibility issues, added the adapters, and fixed the quote issues for consistency. Let me know if you see anything else that needs addressed.

blackbaud-sky-savage commented:
**Tests failed.** Automated cross-browser testing via BrowserStack and Travis CI shows that the JavaScript changes in this pull request are: **BUSTED**

Commit: 14f9e12b0a5e283d86a3620415574ecb63fe2412
Build details: https://travis-ci.org/blackbaud-sky-savage/skyux2/builds/213908589

(*Please note that this is a [fully automated](https://github.com/twbs/savage) comment.*)

Blackbaud-TrevorBurch commented:
Sigh. Missed a selector change. Just pushed the change

Blackbaud-PatrickOFriel commented:
All good @Blackbaud-TrevorBurch , everything else looks good to me, I'll merge it in after the tests pass

blackbaud-sky-savage commented:
**Tests failed.** Automated cross-browser testing via BrowserStack and Travis CI shows that the JavaScript changes in this pull request are: **BUSTED**

Commit: cf50abf8e5d8e7f5a5916a0594c7c59cf7d74de0
Build details: https://travis-ci.org/blackbaud-sky-savage/skyux2/builds/213921047

(*Please note that this is a [fully automated](https://github.com/twbs/savage) comment.*)

Blackbaud-PatrickOFriel commented:
@blackbaud-sky-savage retry

blackbaud-sky-savage commented:
**Tests passed.** Automated cross-browser testing via BrowserStack and Travis CI shows that the JavaScript changes in this pull request are: **CONFIRMED**

Commit: cf50abf8e5d8e7f5a5916a0594c7c59cf7d74de0
Build details: https://travis-ci.org/blackbaud-sky-savage/skyux2/builds/213932967

(*Please note that this is a [fully automated](https://github.com/twbs/savage) comment.*)
